ping/ping.go

```diff
@@ -0,0 +1,212 @@
+	if testnet {
+		Node = SelectedNode{
+			P2PoolStratum: networks.Active.P2ProxyStratum,
+			P2PoolURL:     networks.Active.P2ProxyURL,
+		}
+	} else {
+		Node = SelectedNode{
+			P2PoolStratum: getClosestNodeStratum(),
+			P2PoolURL:     getClosestNodeURL(),
+		}
+	}
+}
+
+var selectedNode string = selector()
+
+func selector() (selected string) {
+	_, err := getNodeInformation("127.0.0.1:9171")
+	if err == nil {
+		logging.Infof("Selected local p2pool node\n")
+		selected = "127.0.0.1:9171"
+		return selected
+	} else {
+		logging.Infof("No local node detected, selecting other public nodes\n")
+	}
+
+	err = pingNodes()
+	if err != nil {
+		logging.Warnf("Nodes could not be pinged, selecting random node")
+		rand.Seed(time.Now().Unix())
+		selected = nodes[rand.Intn(len(nodes))]
+		logging.Infof("%s has been randomly selected", selected)
+		return selected
+	}
+	SortedPingSlice := SortPingtimes()
+
+	for i := 0; i < len(SortedPingSlice); i++ {
+		NodeInformation, _ := getNodeInformation(NodeResults[SortedPingSlice[i]])
+		Fee := CheckFee(NodeInformation)
+		if Fee {
+			CurrentMiners := CheckCurrentMiners(NodeInformation)
+			if CurrentMiners {
+				selected = NodeResults[SortedPingSlice[i]]
+				logging.Infof("%s selected, fulfilled all requirements\n", NodeResults[SortedPingSlice[i]])
+				break
+			} else { //If the node fulfills the max fee requirement but there is more than the set MaxNextping to the next node it will select the current
+				DetermineNextPingTime := SortedPingSlice[i+1]
+				DetermineNextPingTime -= SortedPingSlice[i]
+				if DetermineNextPingTime > Set.MaxNextPing {
+					selected = NodeResults[SortedPingSlice[i]]
+					logging.Infof("%s selected, next node has too high ping time\n", NodeResults[SortedPingSlice[i]])
+					break
+				}
+				logging.Warnf("%s had more than %v miners, trying new inorder to retain efficiency\n", SortedPingSlice[i], Set.MaxMiners)
+			}
+		} else {
+			logging.Warnf("%s had more than a %f fee, trying new\n", SortedPingSlice[i], Set.MaxFee)
+		}
+	}
+	return selected
+}
+
+func pingNodes() error {
+	results := make([]time.Duration, len(nodes))
+
+	for i := 0; i < len(nodes); i++ {
+		pinger, err := ping.NewPinger(nodes[i])
+		pinger.SetPrivileged(true)       //This line is needed for windows because of ICMP
+		pinger.Timeout = Set.PingTimeout //Sets the time for which the pinger will timeout regardless of how many packets there has been recieved
+		if err != nil {
+			logging.Warn("Error: Check if you are connected to the internet")
+			logging.Warn(err)
+			return err
+		}
+		pinger.Count = Set.PingPackets //Number of packets to be sent to each node
+		err = pinger.Run()
+		if err != nil {
+			logging.Warn("Error: Check if you are connected to the internet")
+			logging.Warn(err)
+			return err
+		}
+		if nodes[i] == "p2proxy.vertcoin.org" { //Currently this node uses port 9172 instead of 9171, if this changes this statement can be removed and 9171 can be added to all nodes
```

Comment:
What happens if `pingNodes()` is called again - it seems that you'd end up with endless `:9171` port number suffixes?

Reply:
It would, will change this in the future. Although it's not a problem since the function will only be called once on startup.